### PR TITLE
DOC RFC-8996 clarify public APIs

### PR DESCRIPTION
**Pull request by dnsl48**
https://github.com/silverstripe/silverstripe-framework/issues/8996

### CODE REVIEW

File: docs/en/05_Contributing/01_Code.md [hunk 1]
```diff
@@ -86,6 +86,20 @@ If you are changing existing APIs, introducing new APIs or major new features,
 please review our guidance on [supported versions](release_process#supported-versions).
 You will need to change the branch for your pull request accordingly.
 
+SilverStripe public APIs explicitly include:
+ - namespaces, classes, interfaces and traits
+ - public and protected scope (including methods, properties and constants)
+ - global functions, variables
+ - yml configuration file structure
```
Comment (ScopeyNZ) [inline docs/en/05_Contributing/01_Code.md]: As long as this also includes `yml` variable typing. You shouldn't be able to change an `int` config setting to `int|string` in a patch release, for example.

File: docs/en/05_Contributing/01_Code.md [hunk 2]
```diff
@@ -86,6 +86,20 @@ If you are changing existing APIs, introducing new APIs or major new features,
...
+ - private static class properties (considered to be configuration variables)
+
+SilverStripe public APIs explicitly exclude:
+ - private scope (methods and properties with the exception for `private static`)
+ - entities marked as `@internal`
+ - yml configuration file default values
```
Comment (ScopeyNZ) [inline docs/en/05_Contributing/01_Code.md]: This feels risky to me, but I can't think of any example this would break something.

Comment (dnsl48) [reply docs/en/05_Contributing/01_Code.md]: Doesn't mean we have to change them every patch release. Simply meant to leave it for contributors' discretion

File: docs/en/05_Contributing/01_Code.md (hunk 1 above)
Comment (dnsl48) [reply docs/en/05_Contributing/01_Code.md]: fixed

File: docs/en/05_Contributing/01_Code.md [hunk 3]
```diff
@@ -86,6 +86,20 @@ If you are changing existing APIs, introducing new APIs or major new features,
 please review our guidance on [supported versions](release_process#supported-versions).
 You will need to change the branch for your pull request accordingly.
 
+SilverStripe public APIs explicitly include:
```
Comment (chillu) [inline docs/en/05_Contributing/01_Code.md]: Can we please limit this to PHP APIs? No explicit guarantees about JS APIs. We'll try to keep them stable, but the reality of wrangling a large CMS UI built over a decade is that there's too many moving parts for API stability - at least until we've got this mess under control again through a ReactJS refactor (where we'll create a different kind of mess over time heh)

Comment (dnsl48) [reply docs/en/05_Contributing/01_Code.md]: added JS and everything non-php to explicitly excluded

**Pull request merged**